cpp/plugins/cucim.kit.cuslide/src/cuslide/raw/raw.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Apache License, Version 2.0
+ * Copyright 2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Code below is using libdeflate library which is under MIT license
+ * Please see LICENSE-3rdparty.md for the detail.
+ */
+
+#include "raw.h"
+
+#include <cstring>
+#include <stdexcept>
+#include <unistd.h>
+
+
+namespace cuslide::raw
+{
+
+bool decode_raw(int fd,
+                unsigned char* raw_buf,
+                uint64_t offset,
+                uint64_t size,
+                uint8_t** dest,
```

Comment:
Why does this need to be `**` as opposed to `*`?

Reply:
For the same reason that cudaMalloc accept '**' (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g37d37965bfb4803b6d4e59ff26856356) -- decoders may need to allocate memory and setting the address to `dest` if needed.


See https://github.com/rapidsai/cucim/blob/048d53e6f99c2129b9febd08e0ae6d1b37d74451/cpp/plugins/cucim.kit.cuslide/src/cuslide/jpeg/libjpeg_turbo.cpp#L145

Reply:
Was mostly thinking about this in terms of the `nullptr` checks below. We may want to look into using references for this kind of thing in the future, but not urgent. Can be addressed later